### PR TITLE
Share the majority of code between the Binary and Flat-Array versions of our interval trees.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Use generics and witness types heavily to get entirely specialized codegen for each type.

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/IntervalTreeHelpers.cs [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Shared.Collections;
+
+internal interface IIntervalTreeHelper<T, TIntervalTree, TNode>
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/IntervalTreeHelpers.cs:13]: renamed to IIntervalTreeWitness in followup pr as this is a witness type pattern.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/IntervalTreeHelpers.cs [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        static IEnumerator<T> GetEnumeratorWorker(
+            TIntervalTree tree, TIntervalTreeHelper helper, TNode root)
+        {
+            using var _ = s_nodeStackPool.GetPooledObject(out var stack);
+            var current = (Node: root, HasValue: true);
+
+            while (current.HasValue || stack.Count > 0)
+            {
+                // Traverse all the way down the left side of the tree, pushing nodes onto the stack as we go.
+                while (current.HasValue)
```
Comment (ToddGrun) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/IntervalTreeHelpers.cs:50]: >while (current.HasValue) [](http://example.com/codeflow?start=16&length=24)

Out of curiosity, why the change from the previous GetEnumerator impl?

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/IntervalTreeHelpers.cs:50]: I didn't like the need for the tuple. Or the somewhat non-standard in-order walk vs the normal literature on the topic. 

This also let us move to a single pool of stacks :-)

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/IntervalTreeHelpers.cs [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        if (!helper.TryGetRoot(tree, out var root))
+            return SpecializedCollections.EmptyEnumerator<T>();
+
+        return GetEnumeratorWorker(tree, helper, root);
+
+        static IEnumerator<T> GetEnumeratorWorker(
+            TIntervalTree tree, TIntervalTreeHelper helper, TNode root)
+        {
+            using var _ = s_nodeStackPool.GetPooledObject(out var stack);
+            var current = (Node: root, HasValue: true);
```
Comment (ToddGrun) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/IntervalTreeHelpers.cs:45]: >var current = (Node: root, HasValue: true); [](http://example.com/codeflow?start=12&length=43)

can you expand this out to two separate variables, rather than the valuetuple?

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/IntervalTreeHelpers.cs:45]: sure.

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/IntervalTreeHelpers.cs]: done!

**Review by ToddGrun — APPROVED** [review]
:shipit:
